mule-extensions-api-persistence/src/main/java/org/mule/extension/api/persistence/ModelPropertyMapTypeAdapter.java

```
@@ -0,0 +1,131 @@
+        {
+            if (modelProperty.getValue().isExternalizable())
+            {
+                out.name(getSerializableModelPropertyName(modelProperty.getKey()));
+                final TypeAdapter adapter = gson.getAdapter(modelProperty.getKey());
+                adapter.write(out, modelProperty.getValue());
+            }
+        }
+        out.endObject();
+
+    }
+
+    @Override
+    public Map<Class<? extends ModelProperty>, ModelProperty> read(JsonReader in) throws IOException
+    {
+        final HashMap<Class<? extends ModelProperty>, ModelProperty> modelPropertyHashMap = new HashMap<>();
+
+        in.beginObject();
+        while (in.hasNext())
+        {
+            final Optional<Class<? extends ModelProperty>> modelPropertyClass = getClassForModelProperty(in.nextName());
+
+            if (modelPropertyClass.isPresent())
+            {
+                final Class<? extends ModelProperty> type = modelPropertyClass.get();
+                final TypeAdapter<?> adapter = gson.getAdapter(type);
+                final ModelProperty read = (ModelProperty) adapter.read(in);
+                modelPropertyHashMap.put(type, read);
+            }
+        }
+        in.endObject();
+        return modelPropertyHashMap;
+    }
+
+    private Optional<Class<? extends ModelProperty>> getClassForModelProperty(String modelPropertyName)
+    {
+        Class<? extends ModelProperty> modelPropertyClass = null;
+        if (nameClassMapping.containsKey(modelPropertyName))
+        {
+            modelPropertyClass = nameClassMapping.get(modelPropertyName);
+        }
+        else
+        {
+            try
+            {
+                modelPropertyClass = (Class<? extends ModelProperty>) Class.forName(modelPropertyName);
+            }
+            catch (ClassNotFoundException e)
+            {
+                LOGGER.error("Error loading ModelProperty. Class not found in the current classloader", e);
```

deserealization should fail rather than just logging an error. Throw an exception here